From a7f2d9b2b05ae5e702a7664b180d20f1be163b1b Mon Sep 17 00:00:00 2001
From: Robert Mader <robert.mader@collabora.com>
Date: Thu, 11 Sep 2025 13:04:42 +0200
Subject: [PATCH] wayland: Invert flipped transforms

So the visual results match when running with and without
GDK_DISABLE=offload. The comment above already hinted in this direction,
without being matched by the code - it looks like GdkDihedral applies
flips and rotations in the inverse manner.

This came up when testing the upcoming versions of Snapshot, GTK and the
gtk4paintablesink on mobile phones using the libcamera software ISP. In
this scenario there are usually two cameras, both 90 or 270 degree
rotated - and the front one will get an additional flip in order to
create a "mirror" effect. The recent improvements to the mentioned
components resulted in offloading now being used by default (the
software ISP generates RGB buffers - thus unlike for YUV formats
GDK_DEBUG=color-mgmt is not required for offloading).

Part-of: <https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/8931>
---
 gdk/wayland/gdksubsurface-wayland.c | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/gdk/wayland/gdksubsurface-wayland.c b/gdk/wayland/gdksubsurface-wayland.c
index b49dd9fd0b7..b8f5f98ef26 100644
--- a/gdk/wayland/gdksubsurface-wayland.c
+++ b/gdk/wayland/gdksubsurface-wayland.c
@@ -47,13 +47,29 @@
 static inline enum wl_output_transform
 gdk_texture_transform_to_wl (GdkDihedral transform)
 {
-  return (enum wl_output_transform) transform;
+  enum wl_output_transform tf;
+
+  if (transform == GDK_DIHEDRAL_FLIPPED_90)
+    tf = WL_OUTPUT_TRANSFORM_FLIPPED_270;
+  else if (transform == GDK_DIHEDRAL_FLIPPED_270)
+    tf = WL_OUTPUT_TRANSFORM_FLIPPED_90;
+  else
+    tf = (enum wl_output_transform) transform;
+  return tf;
 }
 
 static inline GdkDihedral
 wl_output_transform_to_gdk (enum wl_output_transform transform)
 {
-  return (GdkDihedral) transform;
+  GdkDihedral tf;
+
+  if (transform == WL_OUTPUT_TRANSFORM_FLIPPED_90)
+    tf = GDK_DIHEDRAL_FLIPPED_270;
+  else if (transform == WL_OUTPUT_TRANSFORM_FLIPPED_270)
+    tf = GDK_DIHEDRAL_FLIPPED_90;
+  else
+    tf = (GdkDihedral) transform;
+  return tf;
 }
 
 G_STATIC_ASSERT ((int) WL_OUTPUT_TRANSFORM_NORMAL == (int) GDK_DIHEDRAL_NORMAL);
-- 
GitLab

